op/model.py

```diff
@@ -302,3 +318,12 @@ def relation_set(self, relation_id, key, value):
 
     def config_get(self):
         return self._run('config-get')
+
+    def is_leader(self):
+        try:
+            self._run('is-leader')
+        except CalledProcessError as e:
+            if b'leadership status unknown' in e.stderr:
+                raise LeadershipStatusUnknown() from e
+            else:
+                raise
```

Comment:
If the process errors, self._run should be able to emit a nice error message for debugging purposes. It's also bad practice to look into error messages from tools, because if the error message changes trivially, the code breaks down silently (we'll never know it changed before someone complains).

In the case at hand, we're not even benefit form the error check, since the outer is_leader is already doing its job, I believe?

Reply:
Then we should change it for `relation_get` and `relation_set` above as well as they look into strerr output (I was following the existing pattern).

The error check is needed in my view because `is-leader` does not accept any arguments that would specify a unit name and "leadership status unknown" is returned by Juju in case some other error condition occurs:

https://github.com/juju/juju/blob/86f470fb6e7d5e706b284f33391ba0f496adf138/worker/uniter/runner/context/leader.go#L67
https://github.com/juju/juju/blob/86f470fb6e7d5e706b284f33391ba0f496adf138/worker/uniter/runner/jujuc/is-leader.go#L47-L51

Reply:
I did comment on that PR as well. IIRC the idea is to actually get juju fixed for the error cases we do care about. This doesn't look like one of them.

> The error check is needed in my view because is-leader does not accept any arguments that would specify a unit name and "leadership status unknown" is returned by Juju in case some other error condition occurs:

We can let the error boil up from _run in a polished form instead of having every single tool call look after every single general error that tools could possibly return. Can you imagine how the logic would look like long term if we keep doing and parse every single error message to raise a specific Python error for all of them?  Can you imagine the life of someone having to maintain that? :)